lib/lte_link_control/lte_lc.c

```diff
@@ -151,32 +182,327 @@ static const char cgauth[] = "AT+CGAUTH="CONFIG_LTE_PDN_AUTH;
+		if (err) {
+			LOG_ERR("Can't parse signalling mode, error: %d", err);
+			return;
+		}
+
+		evt.type = LTE_LC_EVT_RRC_UPDATE;
+		notify = true;
+
+		break;
+	case LTE_LC_NOTIF_CEDRXP:
+		LOG_DBG("+CEDRXP notification");
+
+		err = parse_edrx(response, &evt.edrx_cfg);
+		if (err) {
+			LOG_ERR("Can't parse eDRX, error: %d", err);
+			return;
+		}
+
+		evt.type = LTE_LC_EVT_EDRX_UPDATE;
+		notify = true;
+
+		break;
+	default:
+		LOG_ERR("Unrecognized notification type: %d", notif_type);
+		break;
+	}
+
+	if (evt_handler && notify) {
+		evt_handler(&evt);
 	}
 }
 
+static int parse_psm_cfg(struct at_param_list *at_params,
+			 bool is_notif,
+			 struct lte_lc_psm_cfg *psm_cfg)
+{
+	int err;
+	size_t tau_idx = is_notif ? AT_CEREG_TAU_INDEX :
+				    AT_CEREG_READ_TAU_INDEX;
+	size_t active_time_idx = is_notif ? AT_CEREG_ACTIVE_TIME_INDEX :
+					    AT_CEREG_READ_ACTIVE_TIME_INDEX;
+	char timer_str[9] = {0};
+	char unit_str[4] = {0};
+	size_t timer_str_len = sizeof(timer_str) - 1;
+	size_t unit_str_len = sizeof(unit_str) - 1;
+	size_t lut_idx;
+	u32_t timer_unit, timer_value;
+
+	/* Lookup table for T3324 timer used for PSM active time in seconds.
+	 * Ref: GPRS Timer 2 IE in 3GPP TS 24.008 Table 10.5.163/3GPP TS 24.008.
+	 */
+	static const u32_t t3324_lookup[8] = {2, 60, 600, 60, 60, 60, 60, 0};
+
+	/* Lookup table for T3412 timer used for periodic TAU. Unit is seconds.
+	 * Ref: GPRS Timer 3 in 3GPP TS 24.008 Table 10.5.163a/3GPP TS 24.008.
+	 */
+	static const u32_t t3412_lookup[8] = {600, 3600, 36000, 2, 30, 60,
+					      1152000, 0};
+
+	/* Parse periodic TAU string */
+	err = at_params_string_get(at_params,
+				   tau_idx,
+				   timer_str,
+				   &timer_str_len);
+	if (err) {
+		LOG_ERR("Could not get TAU, error: %d", err);
+		return err;
+	}
+
+	memcpy(unit_str, timer_str, unit_str_len);
+
+	lut_idx = strtoul(unit_str, NULL, 2);
+	if (lut_idx > (ARRAY_SIZE(t3412_lookup) - 1)) {
+		LOG_ERR("Unable to parse periodic TAU string");
+		err = -EINVAL;
+		return err;
+	}
+
+	timer_unit = t3412_lookup[lut_idx];
+	timer_value = strtoul(timer_str + unit_str_len, NULL, 2);
+	psm_cfg->tau = timer_unit ? timer_unit * timer_value : -1;
+
+	/* Parse active time string */
+	err = at_params_string_get(at_params,
+				   active_time_idx,
+				   timer_str,
+				   &timer_str_len);
+	if (err) {
+		LOG_ERR("Could not get TAU, error: %d", err);
+		return err;
+	}
+
+	memcpy(unit_str, timer_str, unit_str_len);
+
+	lut_idx = strtoul(unit_str, NULL, 2);
+	if (lut_idx > (ARRAY_SIZE(t3324_lookup) - 1)) {
+		LOG_ERR("Unable to parse active time string");
+		err = -EINVAL;
+		return err;
+	}
+
+	timer_unit = t3324_lookup[lut_idx];
+	timer_value = strtoul(timer_str + unit_str_len, NULL, 2);
+	psm_cfg->active_time = timer_unit ? timer_unit * timer_value : -1;
+
+	LOG_DBG("TAU: %d sec, active time: %d sec\n",
+		psm_cfg->tau, psm_cfg->active_time);
+
+	return 0;
+}
+
 static int w_lte_lc_init(void)
 {
+	int err;
+
+	if (is_initialized) {
+		return -EALREADY;
```

Comment:
Nit: `EALREADY` and `EINPROGRESS ` are error codes used for non-blocking socket. You should maybe choose another generic error code because this is not a socket-related error.

Reply:
You have used `EPERM` below, I think it makes more sense.